### PR TITLE
feat(studio): MDX runtime engine, two-way component authoring, media pipeline, and studio hardening

### PR DESCRIPTION
**Pull request by itsTarun**
## Summary

This is a major feature branch that replaces the legacy Markdown-only preview with a full MDX runtime engine, adds two-way component authoring with schema-driven insertion, builds a complete media upload pipeline, implements project settings with cascading deletion, and hardens the entire Studio editing experience.

---

### 1. MDX Runtime Engine (replaces react-markdown)

- **Full MDX compilation pipeline** — Client-side `@mdx-js/mdx` compilation with `compileMdx.ts` → `evaluateMdx.ts` → `PreviewRuntime.tsx` replacing the old `react-markdown + rehype-raw` path entirely
- **Dynamic framework adapters** — Per-project adapter system (`lib/repopress/`) that transpiles repo-specific config files (Fumadocs, Astro, Hugo, Docusaurus, Jekyll, etc.) into runtime preview adapters with component mappings and remark/rehype plugins
- **Import rewriting & allowlist registry** — `transformImports.ts` rewrites ESM imports to use an allowlisted module registry, preventing arbitrary code execution while supporting framework components
- **Error boundary UX** — `ErrorBoundary.tsx` catches compilation and runtime errors with user-friendly diagnostics, auto-recovery, and retry controls
- **Preview status indicators** — `PreviewStatus.tsx` shows compilation state, warnings, and adapter diagnostics inline

### 2. Two-Way Component Authoring (Phase 2)

- **Component registry & catalog** — `lib/studio/component-registry.ts` and `component-catalog.ts` provide a typed registry of available components with prop schemas, categories, and documentation
- **ComponentNode intermediate model** — `lib/studio/component-node.ts` represents JSX elements as a structured tree for manipulation, with full prop type support
- **Deterministic MDX serializer** — `lib/studio/component-serializer.ts` converts ComponentNode trees back to valid MDX source with consistent formatting
- **Schema-driven insert UX** — `component-insert-modal.tsx` + `component-prop-form.tsx` let users browse available components, configure props via auto-generated forms, and insert valid JSX into the editor
- **Repo JSX bridge** — `repo-jsx-bridge.tsx` maps repo-specific components to editor descriptors so the MDXEditor recognizes and renders custom JSX inline
- **Auto-discovery fallback** — Editor scans content for `<PascalCase>` tags and creates generic descriptors for components not in the registry

### 3. Media Upload Pipeline

- **Blob-primary upload** — `app/api/media/upload/route.ts` handles multipart file uploads with MIME validation, size limits, and content-type checking
- **GitHub fallback** — Falls back to GitHub blob API for repos without blob storage configured
- **Media resolve API** — `app/api/media/resolve/route.ts` resolves relative image/asset paths to servable URLs with support for private repos (authenticated GitHub raw URLs)
- **Studio integration** — `image-field.tsx` and editor toolbar support drag-and-drop and click-to-upload with live preview of uploaded assets
- **Convex media tracking** — `convex/mediaOps.ts` tracks uploaded assets with metadata (dimensions, MIME type, usage references)

### 4. Project Settings & Cascading Deletion

- **Settings page** — `app/dashboard/[owner]/[repo]/settings/page.tsx` with project configuration and a Danger Zone section
- **Cascading deletion** — `convex/projects.ts` `removeFull` mutation deletes across all 11 related tables (documents, documentHistory, collections, authors, tags, categories, folderMeta, mediaAssets, webhooks, explorerOps, publishBranches) in a single transaction
- **Confirmation UX** — `delete-project-zone.tsx` requires typing the repo name to confirm, preventing accidental deletion
- **Schema indexes** — Added missing indexes for efficient cascading queries

### 5. Studio Editor Hardening

- **Mirror architecture** — `forward-ref-editor.tsx` wraps MDXEditor with proper ref forwarding, controlled updates, and content synchronization
- **Scroll sync** — Content-aware scroll synchronization between editor and preview using `data-scroll-sync-root` markers and proportional mapping
- **Editor-preview split** — Resizable split pane with persistent layout, mobile-responsive collapse, and keyboard shortcuts
- **History compare & restore** — `history-client.tsx` with side-by-side diff viewer (`diff-viewer.tsx`), version restore with auto-transition back to draft status
- **Document state machine** — `convex/documents.ts` enforces the draft→in_review→approved→published lifecycle with ownership checks via `resolveCallerUserId`
- **Tag input** — Fixed dropdown positioning, keyboard navigation, and selection behavior in `tag-input.tsx`
- **Publish operations** — `publish-ops-bar.tsx` and `publish-dialog.tsx` with batch commit support via GitHub Tree/Blob API (`lib/github.ts`)
- **Studio context** — Extended `studio-context.tsx` with adapter state, component schema, and media configuration

### 6. Runtime Hardening (Tasks 1-2 of 6)

- **Persistent adapter cache** — `lib/repopress/adapter-cache.ts` uses IndexedDB with SHA-based invalidation, TTL expiry, and memory fallback
- **GitHub API rate limiting** — `lib/repopress/github-request-control.ts` adds 500ms debounce, sliding-window rate limiting, exponential backoff with jitter, and in-flight request deduplication
- **Frontmatter-aware components** — `PreviewRuntime.tsx` now passes frontmatter to custom components like `DynamicImage` for context-dependent rendering

### 7. Convex Backend Changes

- **Schema additions** — New tables and fields for media assets, explorer operations, publish branches; updated indexes across 10 Convex files
- **Auth hardening** — `resolveCallerUserId()` helper in both `documents.ts` and `documentHistory.ts` unifies auth checks with explicit userId fallback
- **Restore flow** — `documentHistory-restore.ts` extracted restore logic; auto-transitions published documents back to draft on version restore
- **Webhook improvements** — `githubWebhook.ts` with better error handling and event filtering

---

## Test Coverage

**16 test files, 144 tests — all passing**

| Test Suite | Tests | Coverage Area |
|---|---|---|
| `component-registry.test.ts` | 27 | Component registry CRUD, search, validation |
| `component-node.test.ts` | 24 | ComponentNode tree operations, prop handling |
| `component-serializer.test.ts` | 23 | MDX serialization roundtrips, edge cases |
| `integration-insert-pipeline.test.ts` | 18 | End-to-end insert flow: registry → node → serialize |
| `github-request-control.test.ts` | 6 | Rate limiting, debounce, dedup, backoff |
| `media-resolve.test.ts` | 9 | Asset URL resolution, private repo handling |
| `route.test.ts` (media resolve) | 7 | API route validation, auth, error responses |
| `route.test.ts` (media upload) | 6 | Upload validation, MIME checks, size limits |
| `component-catalog.test.ts` | 7 | Catalog queries, filtering, categories |
| `adapter-cache.test.ts` | 3 | IndexedDB cache, TTL, invalidation |
| `component-insert-operation.test.ts` | 3 | Insert operation cursor placement |
| `safe-jsx-prop-eval.test.ts` | 3 | Prop expression safety checks |
| `editor-error-boundary-utils.test.ts` | 3 | Error boundary reset utilities |
| `document-history-restore.test.ts` | 2 | Restore mutation payload generation |
| `github.batch-commit.test.ts` | 2 | Batch commit via Tree/Blob API |
| `compileMdx.test.ts` | 1 | MDX compilation pipeline |

All tests verified via `npx vitest run` and validated during development using Chrome DevTools MCP for live browser testing.

---

## Files Changed

- **177 files** changed across the codebase
- **~18,600 lines added**, **~2,400 lines removed**
- **123 code files** (`.ts`, `.tsx`, `.css`)
- **16 test files** with **2,522 lines** of test code
- **10 Convex backend files** with schema, mutation, and query changes
- **9 plan/documentation files** tracking implementation progress

---

## Remaining Work (Next Phase)

The following items are tracked in `docs/plans/` and are **out of scope for this PR**:

### MDX Runtime Hardening (4 of 6 tasks remaining)
- Task 3: Private asset URL signing for private repos
- Task 4: Expression sandbox (replace `new Function()` with allowlist evaluator) — **security priority**
- Task 5: Compile cache with SHA-based memoization
- Task 6: Plugin merge determinism

### Automation Backbone & Daily Flow (0 of 12 tasks started)
- Scheduled publishing via Convex crons
- Background GitHub credential resolution
- Daily draft seeding pipeline
- Review queue and failure queue UIs
- Full plan at `docs/plans/2026-03-03-automation-backbone-and-daily-flow-plan.md`

---

## How to Test

1. `npm install && npx convex dev` (in one terminal)
2. `npm run dev` (in another terminal)
3. Navigate to `/dashboard`, connect a GitHub repo
4. Open Studio editor — verify MDX preview renders with framework-specific components
5. Insert a component via the toolbar — verify prop form and MDX insertion
6. Upload an image — verify it appears in preview
7. Save draft, publish, and check history restore
8. Run `npx vitest run` — all 144 tests green

### CODE REVIEW

**Comment by itsTarun** [comment]
All review points from the latest pass are now fixed and pushed.

Fix commits:
- 23a1766
- 283fa7d

Addressed items

1. Require auth identity in mediaOps ownership checks (P1)
- Updated `convex/mediaOps.ts` to validate authenticated identity via `authComponent.safeGetAuthUser(ctx)` before ownership checks.
- `requireProjectOwnership` now rejects unauthenticated or mismatched callers.

2. Preserve current content when restoring history versions (P1)
- Updated restore payload construction to snapshot current document state before overwrite.
- `convex/documentHistory-restore.ts` now accepts `currentBody/currentFrontmatter` and `targetBody/targetFrontmatter` separately.
- `convex/documentHistory.ts` now passes current document content into history insert and target history content into document patch.
- Added follow-up guard for missing current body (`currentBody: document.body ?? ""`).

3. Don’t hide restore for newest history entry (P2)
- Updated `app/dashboard/[owner]/[repo]/history/history-client.tsx` to always render the Restore action for history snapshots (removed index-based suppression).

4. Prevent duplicate scheduling in removeFull (P2)
- Updated `convex/projects.ts` `removeFull` to reject duplicate deletion requests when project is already in deleting state (`[DELETING]`).

5. Block global aliases in MDX sandbox shadowing (P1)
- Updated `components/mdx-runtime/evaluateMdx.ts` blocked globals list to include `window`, `self`, and `globalThis`.
- Added regression coverage in `components/mdx-runtime/__tests__/evaluateMdx.test.ts`.

Additional test hardening included
- `lib/__tests__/document-history-restore.test.ts` updated to assert non-lossy restore semantics.
- Updated two stale tests to match current intended behavior:
  - `lib/repopress/__tests__/github-request-control.test.ts`
  - `app/api/media/upload/__tests__/route.test.ts`

Verification (post-push)
- `pnpm lint` ✅ (passes; existing repo-level warnings remain)
- `pnpm test` ✅ (147/147 passing)
- `pnpm build` ✅
- PR checks: Vercel + Vercel Preview Comments ✅

Re-review result
- Reviewed changed code paths for the reported issues and did a fresh verification run.
- No new findings found in the touched scope.

**Comment by itsTarun** [comment]
# Change summary: Addressed critical ID spoofing, XSS vectors, and authorization bypasses.
The latest changes successfully resolve the remaining critical and high-severity security vulnerabilities from the previous reviews. The authentication flow is now properly locked down, the sandbox mitigations are in place, and race conditions are mitigated.

No issues found. Code looks clean and ready to merge.

**Comment by itsTarun** [comment]
Pushed one more security follow-up in `00db712`.

Addressed in this pass:
- `auth.resolveUserIdByGitHubAccount` no longer acts as a public GitHub-account-id -> user-id oracle
- the lookup now requires a short-lived signed server token tied to the specific GitHub account id
- `getPatAuthUserId()` mints that token server-side before calling Convex, so the existing PAT flow keeps working without exposing the lookup publicly
- added regression coverage for the signed GitHub-account lookup token path

Verification after the fix:
- `pnpm lint`
- `pnpm test`

File: app/api/github/publish-ops/route.ts [hunk 1]
```diff
@@ -29,39 +29,32 @@ export async function POST(request: Request) {
...
-    // ── Fetch project details ──────────────────────────────────────────
     const project = await convex.query(api.projects.get, {
       id: projectId as Id<"projects">,
     })
     if (!project) {
       return NextResponse.json({ error: "Project not found" }, { status: 404 })
     }
 
-    // ── Verify ownership ───────────────────────────────────────────────
     if (!userId || project.userId !== userId) {
```
[the remark this answers is not in the document]
Comment (itsTarun) [reply app/api/github/publish-ops/route.ts]: **Resolved.** `userId` is no longer read from the request body. The route now resolves the acting user server-side via `resolveActingUserId()` (OAuth session) with `getPatAuthUserId()` fallback (PAT). Ownership is verified through `verifyProjectAccess(project, actingUserId)` before any mutation proceeds.

Verified in latest commit on `feat/mdx-runtime` — lines 42-48 of `app/api/github/publish-ops/route.ts`.

File: convex/projects.ts [hunk 2]
```diff
@@ -170,6 +194,100 @@ export const getOrCreate = mutation({
...
+        })
+        syncedProjectIds.push(id)
+      }
+    }
+
+    return syncedProjectIds
+  },
+})
+
 export const update = mutation({
```
[the remark this answers is not in the document]
Comment (itsTarun) [reply convex/projects.ts]: **Resolved.** `syncProjectsFromConfig` already had `verifyCallerIdentity(ctx, args.userId)` at the top of its handler. Additionally, the query uses `by_userId_repo` index scoped to `args.userId`, so it only reads/patches projects belonging to the authenticated caller. No cross-user data access is possible.

Verified in latest commit on `feat/mdx-runtime` — `convex/projects.ts`, `syncProjectsFromConfig` handler.

File: convex/projects.ts [hunk 3]
```diff
@@ -218,3 +337,160 @@ export const remove = mutation({
     await ctx.db.delete(args.id)
```
[the remark this answers is not in the document]
Comment (itsTarun) [reply convex/projects.ts]: **Resolved.** `projects.remove` now:
1. Accepts `userId: v.string()` as a required arg
2. Calls `verifyCallerIdentity(ctx, args.userId)` to validate session
3. Fetches the project and checks `project.userId !== args.userId` before deletion

Matches the ownership pattern used by `removeFull`.

Verified in latest commit on `feat/mdx-runtime` — `convex/projects.ts`, `remove` handler.

File: convex/projects.ts [hunk 4]
```diff
@@ -180,6 +298,7 @@ export const update = mutation({
     detectedFramework: frameworkValidator,
```
[the remark this answers is not in the document]
Comment (itsTarun) [reply convex/projects.ts]: **Resolved.** `projects.update` now:
1. Accepts `userId: v.string()` as a required arg
2. Calls `verifyCallerIdentity(ctx, args.userId)` to validate session
3. Fetches the project and checks `project.userId !== args.userId` before patching

Matches the ownership pattern used by `updateFramework`.

Verified in latest commit on `feat/mdx-runtime` — `convex/projects.ts`, `update` handler.

File: components/studio/studio-context.tsx [hunk 5]
```diff
@@ -1,29 +1,28 @@
 import * as React from "react"
```
[the remark this answers is not in the document]
Comment (itsTarun) [reply components/studio/studio-context.tsx]: **Resolved.** All 7 files now have `"use client"` as the first line:

- `components/studio/studio-context.tsx` — `"use client"`
- `components/studio/file-tree-item.tsx` — `"use client"`
- `components/studio/studio-header.tsx` — `"use client"`
- `components/studio/studio-footer.tsx` — `"use client"`
- `components/studio/hooks/use-studio-file.ts` — `"use client"`
- `components/studio/hooks/use-studio-save.ts` — `"use client"`
- `components/studio/hooks/use-studio-publish.ts` — `"use client"`

Verified each file on latest commit on `feat/mdx-runtime`.

**Comment by itsTarun** [comment]
## Re-Review Complete

I did a thorough second pass across all 177 changed files, covering security, auth, ownership, sandbox safety, and code hygiene. **All 11 original review issues remain resolved**, and the broader codebase checks out clean:

### Security & Auth Audit (beyond original review)

| Area | Status | Detail |
|------|--------|--------|
| `convex/explorerOps.ts` | All 6 mutations guarded | `resolveProjectCaller()` with project ownership |
| `convex/mediaOps.ts` | All 5 mutations guarded | `requireProjectOwnership()` with project ownership |
| `convex/documentHistory.ts` | `restoreVersion` guarded | `resolveProjectCaller()` with project ownership |
| `convex/auth.ts` | `resolveUserIdByGitHubAccount` gated | Signed `lookupToken` (HMAC-SHA256 + TTL) required |
| `app/api/media/upload/route.ts` | Server-side auth + validation | `resolveActingUserId()`, 50MB limit, project access check |
| `app/api/media/resolve/route.ts` | Server-side auth | `verifyProjectAccess()` before serving content |
| `evaluateMdx.ts` | Sandbox hardening | 18 dangerous globals blocked, `withEvalGuard` + `withFunctionConstructorGuard` wrappers |
| `lib/project-access-token.ts` | Secure token flow | HMAC-SHA256 signed, TTL-expiring, proper verification |
| Latest commit (`6440a5c`) | Clean | `getOrCreateInternal` is idempotent `internalMutation` (not client-callable), with test coverage |

### Cleanup Required Before Merge

Three review artifact files are still in the branch and should be deleted before merging into `main`:

| File | Size | Reason |
|------|------|--------|
| `pr_review.md` | 3.2 KB | Development review notes |
| `pr_review_2.md` | 6.3 KB | Second-round review notes |
| `review_diff.patch` | **1 MB** | Full diff patch — should never be committed |

These are easy to remove: `git rm pr_review.md pr_review_2.md review_diff.patch && git commit`.

---

@itsyogesh Please review and merge this PR if it looks good.

File: lib/project-access-token.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  const separatorIndex = token.lastIndexOf(".")
+  if (separatorIndex <= 0) return null
+
+  const serialized = decodeURIComponent(token.slice(0, separatorIndex))
+  const signature = token.slice(separatorIndex + 1)
+  if (!(await verifyValue(serialized, signature))) {
+    return null
+  }
+
+  const payload = JSON.parse(serialized) as ProjectAccessTokenPayload
```
[the remark this answers is not in the document]
Comment (itsTarun) [reply lib/project-access-token.ts]: Fixed in e62c3f7. Both `JSON.parse` calls (in `verifyValue` and `verifyProjectAccessToken`) are now wrapped in try/catch. On parse failure, `verifyValue` returns `null` and `verifyProjectAccessToken` returns `false`.
